src/cmd/services/m3coordinator/downsample/downsample_id.go

```diff
@@ -0,0 +1,330 @@
+	tagPairs       []id.TagPair
+	rollupTagIndex int
+
+	tagEncoder serialize.TagEncoder
+	pool       *rollupIDProviderPool
+}
+
+func newRollupIDProvider(
+	tagEncoder serialize.TagEncoder,
+	pool *rollupIDProviderPool,
+) *rollupIDProvider {
+	return &rollupIDProvider{
+		tagEncoder: tagEncoder,
+		pool:       pool,
+	}
+}
+
+func (p *rollupIDProvider) provide(
+	tagPairs []id.TagPair,
+) ([]byte, error) {
+	p.reset(tagPairs)
+	p.tagEncoder.Reset()
+	if err := p.tagEncoder.Encode(p); err != nil {
+		return nil, err
+	}
+	data, ok := p.tagEncoder.Data()
+	if !ok {
+		return nil, fmt.Errorf("unable to access encoded tags: ok=%v", ok)
+	}
+	// Need to return a copy
+	return append([]byte(nil), data.Bytes()...), nil
+}
+
+func (p *rollupIDProvider) reset(
+	tagPairs []id.TagPair,
+) {
+	p.index = -1
+	p.tagPairs = tagPairs
+	p.rollupTagIndex = -1
+	for idx, pair := range tagPairs {
+		if bytes.Compare(rollupTagName, pair.Name) < 0 {
+			p.rollupTagIndex = idx
+			break
+		}
+	}
+	if p.rollupTagIndex == -1 {
+		p.rollupTagIndex = len(p.tagPairs)
+	}
+}
+
+func (p *rollupIDProvider) length() int {
+	return len(p.tagPairs) + 1
+}
+
+func (p *rollupIDProvider) finalize() {
+	if p.pool != nil {
+		p.pool.Put(p)
+	}
+}
+
+func (p *rollupIDProvider) Next() bool {
+	p.index++
+	return p.index < p.length()
+}
+
+func (p *rollupIDProvider) Current() ident.Tag {
+	idx := p.index
+	if idx == p.rollupTagIndex {
+		return ident.Tag{
+			Name:  ident.BytesID(rollupTagName),
+			Value: ident.BytesID(rollupTagValue),
+		}
+	}
+
+	if idx > p.rollupTagIndex {
+		// Effective index is subtracted by 1
+		idx--
+	}
+
+	return ident.Tag{
+		Name:  ident.BytesID(p.tagPairs[idx].Name),
+		Value: ident.BytesID(p.tagPairs[idx].Value),
+	}
+}
+
+func (p *rollupIDProvider) Err() error {
+	return nil
+}
+
+func (p *rollupIDProvider) Close() {
+	// No-op
+}
+
+func (p *rollupIDProvider) Remaining() int {
+	return p.length() - p.index - 1
+}
+
+func (p *rollupIDProvider) Duplicate() ident.TagIterator {
+	duplicate := new(rollupIDProvider)
+	*duplicate = *p
+	duplicate.reset(p.tagPairs)
+	return duplicate
+}
+
+type rollupIDProviderPool struct {
+	tagEncoderPool serialize.TagEncoderPool
+	pool           pool.ObjectPool
+}
+
+func newRollupIDProviderPool(
+	tagEncoderPool serialize.TagEncoderPool,
+	opts pool.ObjectPoolOptions,
+) *rollupIDProviderPool {
+	return &rollupIDProviderPool{
+		tagEncoderPool: tagEncoderPool,
+		pool:           pool.NewObjectPool(opts),
+	}
+}
+
+func (p *rollupIDProviderPool) Init() {
+	p.pool.Init(func() interface{} {
+		return newRollupIDProvider(p.tagEncoderPool.Get(), p)
+	})
+}
+
+func (p *rollupIDProviderPool) Get() *rollupIDProvider {
+	return p.pool.Get().(*rollupIDProvider)
+}
+
+func (p *rollupIDProviderPool) Put(v *rollupIDProvider) {
+	p.pool.Put(v)
+}
+
+func encodedTagsNameAndTags(
+	id []byte,
+	iterPool *encodedTagsIteratorPool,
+) ([]byte, []byte, error) {
+	// ID is always the encoded tags for downsampling IDs
+	metricTags := id
```

Comment:
The id is passed in and returned without getting used in this method, can we just remove it?

Reply:
it's part of the method required, but I can move that to the lambda calling it perhaps.